syntax = "proto3";
package protocol;
option java_package = "io.CEG.sdk.core.extend.protobuf";

import "common.proto";

message Account{
	string address = 1;
	int64  nonce = 2; //Last transaction seq
	AccountPrivilege priv = 3;
	bytes metadatas_hash = 4;	//metadatas_hash = merklehash(metadatas);
	bytes assets_hash = 5;    //assets_hash = merkelhash(assets)
	Contract contract = 6;
	int64 balance = 7;
}

message AssetKey{
	 string issuer = 1;
	 string code = 2;
	 int32 type = 3;		//0:not limited ammount
}

//payAsset
message Asset{
	 AssetKey	key = 1;
	 int64		amount = 2;
}

message AssetProperty{
	int32 decimal = 1;
	string description = 2;
	int64 max_supply = 3;
	int64 issued_amount = 4;
	int32 fee_percent = 5;
}

message AssetStore{
	AssetKey key = 1; //Required
	int64 amount = 2;
	AssetProperty property = 3;
}

message LedgerHeader
{
	int64	seq 		            = 1;
	bytes	hash 					= 2;
	bytes	previous_hash 			= 3;
	bytes	account_tree_hash		= 4;
	int64   close_time 				= 5;
	bytes   consensus_value_hash    = 6;
	int64	version 			    = 7;
	int64	tx_count 				= 8;
	bytes   validators_hash = 9;	// hash
	bytes	fees_hash = 10;
	string	reserve = 11;
	int64	chain_id = 12;
}

message Ledger{
	LedgerHeader header = 1;
    repeated TransactionEnv transaction_envs = 2;
}

message OperationPayAsset{
	string dest_address = 1;
	Asset asset = 2;
	string input = 3;
}

message OperationTypeThreshold{
	Operation.Type type = 1;
	int64 threshold = 2;
}

message AccountPrivilege {
	int64 master_weight = 1;
	repeated Signer signers = 2;
	AccountThreshold thresholds = 3;
}

message AccountThreshold{
	int64 tx_threshold = 1; //required, [-1,MAX(INT64)] -1: indicates no setting
	repeated OperationTypeThreshold type_thresholds = 2;
}

message OperationIssueAsset{
	string code = 1;
	int64 amount = 2;
}

message OperationPayCoin{
	string dest_address = 1;
	int64 amount = 2;
	string input = 3;
}

message OperationSetSignerWeight{
	int64 master_weight = 1; //required, [-1,MAX(UINT32)] -1: indicates no setting
	repeated Signer signers = 2; //address:weight, if weight is 0, it indicates deleting this signer
}

message OperationLog{
	string topic = 1;
	repeated string datas = 2;
}

message OperationSetPrivilege{
	string master_weight = 1;
	repeated Signer signers = 2;
	string tx_threshold = 3;
	repeated OperationTypeThreshold type_thresholds = 4;
}

message Operation {
	enum Type {
		UNKNOWN = 0;
		CREATE_ACCOUNT 			= 1;
		ISSUE_ASSET 			= 2;
		PAY_ASSET               = 3;
		SET_METADATA			= 4;
		SET_SIGNER_WEIGHT		= 5;
		SET_THRESHOLD			= 6;
		PAY_COIN                = 7;
		LOG						= 8;
		SET_PRIVILEGE			= 9;
	};
	Type type = 1;
	string source_address = 2;
	bytes metadata	= 3;

	//
	OperationCreateAccount		create_account 	   = 4;
	OperationIssueAsset			issue_asset 	   = 5;
	OperationPayAsset			pay_asset 		   = 6;
	OperationSetMetadata		set_metadata	   = 7;
	OperationSetSignerWeight	set_signer_weight  = 8;
	OperationSetThreshold		set_threshold 	   = 9;
	OperationPayCoin			pay_coin           = 10;
	OperationLog				log				   = 11;
	OperationSetPrivilege		set_privilege	   = 12;
}

message OperationSetThreshold{
	int64 tx_threshold = 1;
	repeated OperationTypeThreshold type_thresholds = 2; //type:threshold, threshold:0 indicates deleting this type
}

message Transaction {
	enum Limit{
		UNKNOWN = 0;
		OPERATIONS = 1000;
	};
	string source_address = 1;
	int64 nonce = 2;
	int64  fee_limit = 3;
	int64  gas_price =4;
	int64 ceil_ledger_seq = 5;
	bytes metadata = 6;
	repeated Operation operations = 7;
	int64 chain_id = 8;
}

message Signer {
	enum Limit{
		SIGNER_NONE = 0;
		SIGNER = 100;
	};
	string address = 1;
	int64 weight = 2;
}

message Trigger{
	enum TransactionType{
		NORMAL_TRANSACTION	 = 0;
		CONTRACT_TRANSACTION = 1;
	};

	message OperationTrigger{
		bytes hash = 1;
		int64 index = 2;
	}
	TransactionType transaction_type = 1;
	int64 ledger_seq = 2;  //If it is triggered by a specific block (temporarily not provided)
	OperationTrigger transaction = 3;  //If it's a transaction triggered
	
}

enum Limit{
	UNKNOWN = 0;
	SIGNATURE = 100;
};

message TransactionEnv {
	Transaction transaction = 1;
	repeated Signature signatures 	= 2;
	Trigger trigger = 3;
}

message TransactionEnvStore {
	TransactionEnv transaction_env = 1;
	int32 error_code = 2;
	string error_desc = 3;
	int64 ledger_seq = 4;
	int64 close_time = 5;
	//for notify
	bytes hash = 6;
	int64 actual_fee = 7;
	repeated bytes contract_tx_hashes = 8;
}

message TransactionEnvSet{
	repeated TransactionEnv txs = 2;
}

message ConsensusValueValidation {
	repeated int32 expire_tx_ids = 1;  //Contract execution expired
	repeated int32 error_tx_ids = 2;
	//May contain other checks
}

message ConsensusValue{
	TransactionEnvSet txset = 1;
	int64 close_time = 2;
	bytes previous_proof = 3;
	int64 ledger_seq = 4;
	bytes previous_ledger_hash = 5;
	LedgerUpgrade ledger_upgrade = 6;
	ConsensusValueValidation validation = 7;
}

message Contract{
    enum ContractType{
		JAVASCRIPT = 0;
	}
	ContractType type = 1;
	string payload = 2;
}

message OperationCreateAccount{
	string dest_address = 1;
	Contract contract = 2;
	AccountPrivilege priv = 3;
	repeated KeyPair metadatas = 4;	
	int64	init_balance = 5;
	string  init_input = 6;
}

message OperationSetMetadata{
	string	key = 1;  
	string  value = 2;
	int64 	version = 3; //Optional. The version number of metadata. If set, it must be the same as the existing metadata_version of the account.
	bool    delete_flag = 4;
}
